Specify that all strings can be Unicode. (#3079)
modelica/ModelicaSpecification: chapters/functions.tex

An exception is made when the argument is of the form \lstinline!size($\ldots$, $\ldots$)!. In this case the corresponding C type is \lstinline!size_t!.

> Strings are \textsc{nul}-terminated (i.e., terminated by \lstinline[language=C]!'\0'!) and are encoded using UTF-8 (assuming \lstinline[language=C]!CHAR_BIT==8! in C) to facilitate calling of C functions.

Reviewer: As the most immediate impact of the choice of encoding is on the MSL string handling functions, it is better to keep this choice open until we have figured out what functionality the MSL string handling functions should provide if `String` values were Unicode, and the proper place for discussing this appears to be https://github.com/modelica/ModelicaStandardLibrary/issues/3789.

Author: > I don't think I see what you mean. What duplication? Can we be explicit about which are the other strings that are not of type `String`?

All attributes of StringType, including the start-attribute for String. The duplication is that we have the Unicode-rule twice.

Reviewer: I see, so why not simply specify that `StringType` may hold any Unicode characters?

Reviewer: Details can now be discussed at the added text instead of here.  Resolving.

An exception is made when the argument is of the form \lstinline!size($\ldots$, $\ldots$)!. In this case the corresponding C type is \lstinline!size_t!.

Strings are \textsc{nul}-terminated (i.e., terminated by \lstinline[language=C]!'\0'!) to facilitate calling of C functions.

> Strings may contain Unicode and for 8-bit char in C they are coded as UTF-8.

Reviewer 2: I think I would rather want to combine the 2 sentences into one.

Author: Done now.
I stated that it only applies for C-implementations with 8-bit characters (as on practically all modern computers); I don't see the need to consider other cases yet.

Reviewer: @sjoelund, could you please resolve if you are satisfied with the new sentence.

\end{tabular}
\end{center}

> Sending string literals to FORTRAN~77 subroutines/functions is supported for Lapack/Blas-routines, and the strings are \textsc{nul}-terminated for compatibility with C.

Reviewer: Could a FORTRAN expert shed some light on this?  Why would we need to define null-termination of strings passed to FORTRAN?

Author: Since the NUL-termination wasn't changed I don't see it as relevant for this PR.
If the Fortran-code is transformed into C-code using f2c I can see that it helps; and if the strings already are C-strings there's zero cost.

Reviewer: Right, it isn't relevant for this PR.  I see that there is probably zero cost for all tools, but I wouldn't have expected this to be of any value even when using f2c.  Maybe you can elaborate on this some other time…

Resolving.

> The strings should be UTF-8 encoded similarly as for C to be compatible with C, even though the support in FORTRAN~77 is unclear and it is not relevant for the Lapack/Blas-routines.

Reviewer: I don't see a need to aim for compatibility with C here.  Why not simply this?
```suggestion
Strings are UTF-8 encoded, even though the support for non-ASCII characters in FORTRAN~77 is unclear and it is not relevant for the Lapack/Blas-routines.
```

Author: Done.

Author: Now it is done.

Reviewer: Looks good.  Resolving.

> The strings should be UTF-8 encoded similarly as for C to be compatible with C, even though the support in FORTRAN~77 is unclear and it is not relevant for the Lapack/Blas-routines.

Reviewer 3: ```suggestion
The strings should be UTF-8 encoded similarly as for C to be compatible with C, even though the support in FORTRAN~77 is unclear and it is not relevant for the LAPACK/BLAS-routines.
```
Note that LAPCK no longer is FORTRAN 77, but FORTRAN 90 since v3.2.

Reviewer: Right, so modern LAPACK isn't usable with Modelica.

Reviewer 3: Exactly.
